encoding/mvt/layer.go

```diff
@@ -0,0 +1,192 @@
+package mvt
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+
+	"context"
+
+	vectorTile "github.com/go-spatial/geom/encoding/mvt/vector_tile"
+)
+
+// Layer describes a layer within a tile.
+// Each layer can have multiple features
+type Layer struct {
+	// Name is the unique name of the layer within the tile
+	Name string
+	// The set of features
+	features []Feature
+	// default is 4096
+	extent *int
+}
+
+func valMapToVTileValue(valMap []interface{}) (vt []*vectorTile.Tile_Value) {
+	for _, v := range valMap {
+		vt = append(vt, vectorTileValue(v))
+	}
+
+	return vt
+}
+
+// VTileLayer returns a vectorTile Tile_Layer object that represents this layer.
+func (l *Layer) VTileLayer(ctx context.Context) (*vectorTile.Tile_Layer, error) {
+	kmap, vmap, err := keyvalMapsFromFeatures(l.features)
+	if err != nil {
+		return nil, err
+	}
+
+	valmap := valMapToVTileValue(vmap)
+
+	var features = make([]*vectorTile.Tile_Feature, 0, len(l.features))
+	for _, f := range l.features {
+		// context check
+		if err := ctx.Err(); err != nil {
+			return nil, err
+		}
+
+		vtf, err := f.VTileFeature(ctx, kmap, vmap)
+		if err != nil {
+			switch err {
+			case context.Canceled:
+				return nil, err
+			default:
+				return nil, fmt.Errorf("error getting VTileFeature: %v", err)
+			}
+		}
+
+		if vtf != nil {
+			features = append(features, vtf)
+		}
+	}
+
+	ext := uint32(l.Extent())
+	version := uint32(l.Version())
+	vtl := new(vectorTile.Tile_Layer)
+	vtl.Version = &version
+	name := l.Name // Need to make a copy of the string.
+	vtl.Name = &name
+	vtl.Features = features
+	vtl.Keys = kmap
+	vtl.Values = valmap
+	vtl.Extent = &ext
+
+	return vtl, nil
+}
+
+// Version is the version of tile spec this layer is from.
+func (*Layer) Version() int { return 2 }
+
+// Extent defaults to 4096
+func (l *Layer) Extent() int {
+	if l == nil || l.extent == nil {
+		return 4096
+	}
+	return *(l.extent)
+}
+
+// SetExtent sets the extent value
+func (l *Layer) SetExtent(e int) {
+	if l == nil {
+		l = new(Layer)
+	}
```

Comment:
I don't think if guard is doing what I thought It did. Change it to just return.
e.g.: [here is a play](https://play.golang.org/p/2CRfBySigB_8)